libyara/scan.c

```diff
@@ -999,6 +1001,25 @@ int yr_scan_verify_match(
   if (STRING_IS_DISABLED(string))
     return ERROR_SUCCESS;
 
+  if (context->matches[string->idx].count == YR_MAX_STRING_MATCHES)
+  {
+    result = callback(
+        context,
+        CALLBACK_MSG_SCAN_WARNING_MATCH_LIMIT,
+        (void*) string,
+        context->user_data);
+
+    // Bump the counter so we don't call the callback again.
+    context->matches[string->idx].count++;
```

Comment:
I think this logic is not correct. When the count reaches `YR_MAX_STRING_MATCHES` the callback is called with `CALLBACK_MSG_SCAN_WARNING_MATCH_LIMIT`, let's say that the callback returns `CALLBACK_CONTINUE`, in which case the scanning will continue. If some other match is found for the same string now the count is larger than `YR_MAX_STRING_MATCHES` and the callback won't be called again, so YARA will behave as always a continue tracking the matches.

It looks like with this logic YARA is ignoring a **single** match.

Reply:
You're right. I'll get things fixed up (and add tests that make me more confident I'm doing the right thing) in the next day or so.

Reply:
I think I've fixed this now, and I also added a test case which should keep me honest in the future.

If you think this is right I'll take a shot at adding the documentation for this.